L1Trigger/TrackFindingTMTT/src/Stub.cc

```diff
@@ -0,0 +1,789 @@
+
+    if (trackerGeometryVersion_ == "T5") {
+      if (!barrel_) {
+        // Apply bodge, since Topology class annoyingly starts ring count at 1, even in endcap wheels where
+        // inner rings are absent.
+        unsigned int iWheel = trackerTopology->tidWheel(detId);
+        if (iWheel >= 3 && iWheel <= 5)
+          endcapRing_ += 3;
+      }
+    }
+
+    // Get sensor strip or pixel pitch using innermost sensor of pair.
+
+    const PixelGeomDetUnit* unit = reinterpret_cast<const PixelGeomDetUnit*>(det0);
+    const PixelTopology& topo = unit->specificTopology();
+    const Bounds& bounds = det0->surface().bounds();
+
+    std::pair<float, float> pitch = topo.pitch();
+    stripPitch_ = pitch.first;      // Strip pitch (or pixel pitch along shortest axis)
+    stripLength_ = pitch.second;    //  Strip length (or pixel pitch along longest axis)
+    nStrips_ = topo.nrows();        // No. of strips in sensor
+    sensorWidth_ = bounds.width();  // Width of sensitive region of sensor (= stripPitch * nStrips).
+
+    // Note if modules are flipped back-to-front.
+    outerModuleAtSmallerR_ = (det0->position().mag() > det1->position().mag());
+    /*
+  if ( barrel_ && det0->position().perp() > det1->position().perp() ) {
+    outerModuleAtSmallerR_ = true;
+  }
+  */
+
+    sigmaPerp_ = stripPitch_ / sqrt(12.);  // resolution perpendicular to strip (or to longest pixel axis)
+    sigmaPar_ = stripLength_ / sqrt(12.);  // resolution parallel to strip (or to longest pixel axis)
+  }
+
+  //=== Determine tracker geometry version by counting modules.
+
+  void Stub::setTrackerGeometryVersion(const TrackerGeometry* trackerGeometry, const TrackerTopology* trackerTopology) {
+    // N.B. The relationship between Tracker geometry name (T3, T4, T5 ...) and CMS geometry name
+    // (D13, D17 ...) is documented in
+    //  https://github.com/cms-sw/cmssw/blob/CMSSW_9_1_X/Configuration/Geometry/README.md .
+
+    if (trackerGeometryVersion_ == "UNKNOWN") {
+      unsigned int numDet = 0;
+      for (const GeomDet* gd : trackerGeometry->dets()) {
+        DetId detid = gd->geographicalId();
+        if (detid.subdetId() != StripSubdetector::TOB ||
+            detid.subdetId() !=
+                StripSubdetector::TID) {  // Phase 2 Outer Tracker uses TOB for entire barrel & TID for entire endcap.
+          if (trackerTopology->isLower(detid)) {  // Select only lower of the two sensors in a module.
+            numDet++;
+          }
+        }
+      }
+
+      if (numDet == 13296) {
+        trackerGeometryVersion_ = "T5";  // Tilted geometry
+      } else if (numDet == 13556) {
+        trackerGeometryVersion_ = "T3";  // Older tilted geometry
+      } else if (numDet == 14850) {
+        trackerGeometryVersion_ = "T4";  // Flat geometry
+      } else {
+        trackerGeometryVersion_ = "UNRECOGNISED";
+        cout << "Stub: WARNING -- The tracker geometry you are using is yet not known to the stub class. Please update "
```

Comment:
Not thread safe, please use MessageLogger (or throw an exception?).

Reply:
done